### PR TITLE
feat: republish support  qos & retain placeholder

### PR DESCRIPTION
**Pull request by DDDHuang**

### CODE REVIEW

File: apps/emqx_rule_engine/src/emqx_rule_actions.erl [hunk 1]
```diff
@@ -211,3 +260,54 @@ format_msg([], Data) ->
...
+    get_qos(emqx_rule_utils:replace_simple_var(Tmpl, Data)).
+
+get_qos(<<"-1">>) -> {ok, 0};
+get_qos(<<"0">>) ->  {ok, 0};
+get_qos(<<"1">>) ->  {ok, 1};
+get_qos(<<"2">>) ->  {ok, 2};
+get_qos(0) ->        {ok, 0};
+get_qos(1) ->        {ok, 1};
+get_qos(2) ->        {ok, 2};
+get_qos(_) ->        {error, bad_qos}.
```
Comment (terry-xiaoyu) [inline apps/emqx_rule_engine/src/emqx_rule_actions.erl:287]: `get_qos(_QoS) -> error({bad_qos, _QoS}).`

File: apps/emqx_rule_engine/src/emqx_rule_actions.erl [hunk 2]
```diff
@@ -211,3 +260,54 @@ format_msg([], Data) ->
...
+
+get_retain(true) ->        {ok, true};
+get_retain(false) ->       {ok, false};
+get_retain(<<"true">>) ->  {ok, true};
+get_retain(<<"false">>) -> {ok, false};
+get_retain(<<"1">>) ->     {ok, true};
+get_retain(<<"0">>) ->     {ok, false};
+get_retain(1) ->           {ok, true};
+get_retain(0) ->           {ok, false};
+get_retain(_) ->           {error, bad_retain}.
```
Comment (terry-xiaoyu) [inline apps/emqx_rule_engine/src/emqx_rule_actions.erl:313]: `get_retain(_Retain) -> error({bad_retain, _Retain}).`

File: apps/emqx_rule_engine/test/emqx_rule_engine_SUITE.erl [hunk 3]
```diff
@@ -413,7 +413,8 @@ t_republish_action(_Config) ->
                     #{rawsql => <<"select topic, payload, qos from \"t1\"">>,
                       actions => [#{name => 'republish',
                                     args => #{<<"target_topic">> => <<"t2">>,
-                                              <<"target_qos">> => -1,
+                                              <<"target_qos">> => <<"-1">>,
+                                              <<"target_retain">> => <<"false">>,
```
Comment (terry-xiaoyu) [inline apps/emqx_rule_engine/test/emqx_rule_engine_SUITE.erl:417]: <<"target_qos">> => <<"${qos}">>,

File: apps/emqx_rule_engine/src/emqx_rule_actions.erl [hunk 4]
```diff
@@ -35,24 +35,49 @@
...
-                                 zh => <<"重新发布消息时用的 QoS 级别, 设置为 -1 以使用原消息中的 QoS"/utf8>>}
+                description => #{en =>
+                                    <<"The QoS Level to be uses when republishing the message."
+                                      " Support placeholder variables."
+                                      " Set to ${qos} to use the original QoS."
+                                      " Or other variable, value is 0 or 1 or 2">>,
+                                 zh =>
+                                    <<"重新发布消息时用的 QoS 级别。"
+                                      "支持占位符变量，可以填写 ${qos} 来使用原消息的 QoS，"
+                                      "或其他值为 0 或 1 或 2 的变量。"/utf8>>}
```
Comment (terry-xiaoyu) [inline apps/emqx_rule_engine/src/emqx_rule_actions.erl:53]: improve the descriptions

File: apps/emqx_rule_engine/src/emqx_rule_utils.erl [hunk 5]
```diff
@@ -87,6 +91,22 @@ preproc_tmpl([[Str, Phld]| Tokens], Acc) ->
 preproc_tmpl([[Str]| Tokens], Acc) ->
     preproc_tmpl(Tokens, put_head(str, Str, Acc)).
 
+%% Replace a string contains vars to another string in which the placeholders are replace by the
+%% corresponding values. For example, given "a: ${var}", if the var=1, the result string will be:
+%% "a: 1".
+replace_vars_in_str(Tokens, Data) when is_list(Tokens) ->
```
Comment (savonarola) [inline apps/emqx_rule_engine/src/emqx_rule_utils.erl:97]: Couldn't be the names a bit misleading 🤔 The replacement ocurrs not in a string but in a (maybe) preprocessed template, and if we actually pass a string `replace_vars_in_str("a: ${var}", ...)` probably a clause error will occur somewhere in `proc_tmpl`

**Review by terry-xiaoyu — DISMISSED** [review]

File: apps/emqx_rule_engine/src/emqx_rule_utils.erl (hunk 5 above)
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_utils.erl:97]: So could we change the name to `replace_vars_to_str`? This function is to convert tokens to a string (binary()).

**Review by terry-xiaoyu — DISMISSED** [review]

**Pull request merged**